### PR TITLE
fix(typedarray): close buffer conformance gaps

### LINKED ISSUE #840: Bring TypedArray, ArrayBuffer and DataView to 100% test262 conformance (~56 fails)
## Summary

Drive the test262 **TypedArray, ArrayBuffer and DataView** failures to zero across `built-ins`, `staging`, and `intl402` — these share root causes, so a single fix typically clears the failure in more than one category.

## Why

Reaching ~100% test262 on supported categories is the objective for the conformance milestone. Baseline (main, 2026-06-26): **97.3% overall** (50,877/52,293; 0 timeouts, 0 infra; `language` and `harness` already at 100%). **TypedArray, ArrayBuffer and DataView accounts for ~56** of the 1,416 remaining failures.

## Current behavior

~56 TypedArray, ArrayBuffer and DataView tests fail. Concentrations: TypedArray prototype + element access, ArrayBuffer / SharedArrayBuffer, DataView, Uint8Array base64/hex.

## Expected behavior

Zero TypedArray, ArrayBuffer and DataView failures at the pinned suite SHA (`scripts/test262-suite-sha.txt`), with no regressions elsewhere — the PR conformance gate enforces no PASS to non-PASS transitions.

## Scope notes

- Organize the work by failing sub-area (above); they are independently fixable.
- Land the conformance gates first: harness diagnostics (#830), the crash fix (#831), the skip-list (#832).
- Every fix ships with its JS test in the same commit. Parent #820.

### PR DESCRIPTION
**Pull request by frostney**
## Summary

- Complete the TypedArray, ArrayBuffer, SharedArrayBuffer, DataView, and Uint8Array edge cases needed for the pinned issue #840 conformance slice.
- Align native constructor finalization/prototype lookup ordering across interpreter and bytecode paths, including `Reflect.construct`/`newTarget` cases and delayed ArrayBuffer/DataView allocation checks.
- Add ArrayBuffer immutable/sliceToImmutable behavior, SharedArrayBuffer species slicing, DataView detached/index ordering, TypedArray slice/set semantics, Uint8Array base64/hex decoding edge cases, and missing test262 realm host hooks.
- Follow-up audit fixes: bytecode native subclasses now validate ArrayBuffer/SharedArrayBuffer lengths and DataView offsets before touching `newTarget.prototype`; zero-length `SharedArrayBuffer.prototype.slice()` now returns a distinct empty buffer.
- Final failure sweep: direct bytecode `Object(value)` construction boxes primitives, VM class construction preserves private brand initialization for eval-authored classes, typed-array prototype lookup ordering matches test262, object-pattern shorthand now rejects keyword bindings while allowing explicit keyword property aliases, and function-super derived classes preserve explicit `newTarget` prototype/realm fallback.
- Review cleanup: `$262.createRealm()` now returns the child `$262` host object with mirrored hooks, native implicit constructors keep the allocated native instance rooted through initialization/finalization, and `new this.#Ctor()` parses private constructor member callees.
- Parser nitpick cleanup: `Call()` and `new` callee parsing now share the same member-access segment parser, while object-pattern keyword shorthand still reaches the precise binding-validation error path.
- Closes #840.

## Testing

- [x] Verified no regressions and confirmed the new feature or bugfix in end-to-end JavaScript/TypeScript tests
  - `./build.pas --clean testrunner`
  - `./build.pas --clean loaderbare`
  - `./build/GocciaTestRunner tests` -> 11196/11196
  - `./build/GocciaTestRunner tests --mode=bytecode` -> 11196/11196
  - `./build/GocciaTestRunner tests/built-ins/ArrayBuffer/constructor.js tests/built-ins/SharedArrayBuffer/constructor.js tests/built-ins/DataView/constructor.js tests/built-ins/SharedArrayBuffer/prototype/slice.js tests/language/identifiers/reserved-words-as-properties.js` -> 101/101
  - `./build/GocciaTestRunner tests/built-ins/ArrayBuffer/constructor.js tests/built-ins/SharedArrayBuffer/constructor.js tests/built-ins/DataView/constructor.js tests/built-ins/SharedArrayBuffer/prototype/slice.js tests/language/identifiers/reserved-words-as-properties.js --mode=bytecode` -> 101/101
  - `./build/GocciaTestRunner tests/built-ins/Reflect/construct.js tests/built-ins/Reflect/construct/function-keyword.js tests/built-ins/Reflect/construct/class-constructor-return.js` -> 72/72
  - `./build/GocciaTestRunner tests/built-ins/Reflect/construct.js tests/built-ins/Reflect/construct/function-keyword.js tests/built-ins/Reflect/construct/class-constructor-return.js --mode=bytecode` -> 72/72
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/TypedArray/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-typedarray.json` -> 1446/1446
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/ArrayBuffer/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-arraybuffer.json` -> 221/221
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/SharedArrayBuffer/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-sharedarraybuffer.json` -> 104/104
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/DataView/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-dataview.json` -> 561/561
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/Uint8Array/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-uint8array.json` -> 70/70
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/**/TypedArray/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-staging-typedarray.json` -> 96/96
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/**/ArrayBuffer/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-staging-arraybuffer.json` -> 5/5
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/**/DataView/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-staging-dataview.json` -> 3/3
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/Uint8Array/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-staging-uint8array.json` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories intl402 --filter 'intl402/TypedArray/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-intl402-typedarray.json` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/Object/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-object-after.json` -> 3411/3411
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/String/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-string-after.json` -> 1223/1223
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/TypedArrayConstructors/ctors/typedarray-arg/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-typedarray-arg-post-super-fix.json` -> 14/14
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories language --filter 'language/expressions/class/**' --mode bytecode --jobs 4 --output /tmp/goccia-test262-class-expr-post-super-fix.json` -> 4059/4059
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories language --filter 'language/expressions/super/realm.js' --mode bytecode --jobs 1 --output /tmp/goccia-test262-super-realm-final.json` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/sm/regress/regress-694306.js' --mode bytecode --jobs 1 --output /tmp/goccia-test262-sm-694306-post-super-fix.json` -> 1/1
  - `./format.pas --check`
  - `git diff --check`
  - `./build/GocciaTestRunner tests/language/classes/private-getter-call.js tests/built-ins/Reflect/construct/function-keyword.js tests/built-ins/Reflect/construct/class-constructor-return.js` -> 54/54
  - `./build/GocciaTestRunner tests/language/classes/private-getter-call.js tests/built-ins/Reflect/construct/function-keyword.js tests/built-ins/Reflect/construct/class-constructor-return.js --mode=bytecode` -> 54/54
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/goccia/create-realm-host-hooks.js' --mode bytecode --jobs 1 --output /tmp/goccia-test262-review-create-realm.json` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories staging --filter 'staging/sm/TypedArray/set-wrapped.js' --mode bytecode --jobs 1 --output /tmp/goccia-test262-review-set-wrapped.json` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/ArrayBuffer/options-maxbytelength-compared-before-object-creation.js' --mode bytecode --jobs 1` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/SharedArrayBuffer/options-maxbytelength-compared-before-object-creation.js' --mode bytecode --jobs 1` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/DataView/proto-from-ctor-realm.js' --mode bytecode --jobs 1` -> 1/1
  - `bun scripts/run_test262_suite.ts --suite-dir /tmp/goccia-test262-020cb740 --categories built-ins --filter 'built-ins/DataView/proto-from-ctor-realm-sab.js' --mode bytecode --jobs 1` -> 1/1
- [ ] Updated documentation
  - N/A: runtime conformance/test coverage only; no user-facing docs changed.
- [ ] Optional: Verified no regressions and confirmed the new feature or bugfix in native Pascal tests (if AST, scope, evaluator, or value types changed)
  - Not run separately; covered by full JS runner gates and focused test262 conformance filters.
- [ ] Optional: Verified no benchmark regressions or confirmed benchmark coverage for the change
  - Not run locally; this change targets correctness/conformance rather than a benchmarked hot path.

## Review Evidence

- Post-main-merge validation: `./build.pas --clean testrunner`, full interpreted/bytecode JS suites (11196/11196 each; 24463 assertions), `./format.pas --check`, and `git diff --check` all passed.
- Parser nitpick validation: `./build.pas --clean testrunner`, `./build/GocciaTestRunner tests/language/identifiers/reserved-words-as-properties.js`, full interpreted/bytecode JS suites (11196/11196 each; 24463 assertions), `./format.pas --check`, and `git diff --check` all passed.
- Manual diff review after the follow-up audit covered native subclass construction, bytecode constructor redirects, delayed native receiver initialization, direct `Object(value)` bytecode construction, eval-authored private class construction, object-pattern keyword shorthand validation, function-super `newTarget` prototype/realm fallback, zero-length SharedArrayBuffer slice identity, `$262.createRealm()` host-object shape, native constructor GC rooting, and private-member `new` callees.
- Parallel audit agents identified the earlier native-constructor and SharedArrayBuffer follow-up findings; both now have regression coverage and pass in interpreter and bytecode modes.
- The final failure sweep was validated against pinned test262 checkout `020cb74075849d1e404bbcdb62feb7a02e6966db`.

### CODE REVIEW

File: scripts/test262_harness/$262.js [hunk 1]
```diff
@@ -21,12 +21,34 @@ const AbstractModuleSource = () => {
...
+      return realm.evalScript(sourceText);
+    },
+
+    createRealm() {
+      return installRealm262(realm.createRealm());
+    },
+  };
+  realm.global.$262 = realm262;
+  return realm;
+}
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/test262_harness/$262.js:45]: Fixed in 9e2e6791. installRealm262 now installs and returns the child `` host object, and mirrors AbstractModuleSource/IsHTMLDDA onto child realms.

File: source/units/Goccia.Evaluator.pas [hunk 2]
```diff
@@ -3624,10 +3632,27 @@ function InvokeConstructableWithReceiver(const AConstructor: TGocciaValue;
...
+        ThrowTypeError('Superclass constructor did not return an object',
+          SSuggestNotConstructorType);
+      if NativeInstance is TGocciaInstanceValue then
+        TGocciaInstanceValue(NativeInstance).InitializeNativeFromArguments(AArguments);
+      ReceiverPrototype := GetProtoFromConstructor(EffectiveNewTarget);
+      NativeInstance.Prototype := ReceiverPrototype;
+      if NativeInstance is TGocciaInstanceValue then
+        TGocciaInstanceValue(NativeInstance).FinalizeNativeFromArguments(AArguments);
+      Result := NativeInstance;
+    end
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.pas]: Fixed in 9e2e6791. The native implicit-constructor path now temp-roots the allocated native instance through initialization/prototype lookup/finalization and uses GetNativePrototypeFromConstructor for intrinsic fallback parity.

File: source/units/Goccia.Parser.pas [hunk 3]
```diff
@@ -2643,25 +2643,36 @@ function TGocciaParser.Primary: TGocciaExpression;
...
+              Token := Advance;
+              // Reserved words are valid property names after "." per ES spec
+              Expr := TGocciaMemberExpression.Create(Expr, Token.Lexeme, False,
+                Token.Line, Token.Column, False);
+            end
+            else
+              raise TGocciaSyntaxError.Create('Expected property name after "."', Peek.Line, Peek.Column, FFileName, FSourceLines,
+                SSuggestPropertyNameIdentifier);
+          end;
         end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Parser.pas]: Fixed in 9e2e6791. ParseNewExpression now mirrors the Call() private-member branch for `.#name` callee chains, with a regression covering `new this.#Ctor(...)`.
